fix: Encapsulate `Wcmd` within a private wrapper

surajk-m opened the pull request:
Closes #225 

The change encapsulates `Wcmd` within a private wrapper.

surajk-m committed:
fix: Encapsulate `Wcmd` within a private wrapper

surajk-m committed:
cargo fmt

jonhoo commented on src/session.rs:
```diff
@@ -332,11 +336,8 @@ impl Client {
-        cmd: impl WebDriverCompatibleCommand + Send + 'static,
-    ) -> Result<Json, error::CmdError> {
-        self.issue(Cmd::WebDriver(Box::new(cmd))).await
+    pub async fn issue_cmd(&self, cmd: Wcmd) -> Result<Json, error::CmdError> {
```
I think this specifically needs to _not_ change, otherwise we're putting `Wcmd` back into the public API here.

surajk-m replied:
Apologies for the oversight. I'll revise it to use the `WebDriverCompatibleCommand` trait.

jonhoo reviewed (changes requested):
Thanks for taking this on!

surajk-m committed:
Add `test_wrap_command` to wrap `WebDriverCommand` in tests

jonhoo commented on Cargo.toml:
```diff
@@ -47,6 +48,7 @@ tokio = { version = "1", features = ["full"] }
 hyper = { version = "1.1.0", features = ["server"] }
 hyper-util = { version = "0.1.3", features = ["server", "http1"] }
 serial_test = "3.0"
+fantoccini = { path = "../fantoccini", features = ["test_helpers"] }
```
This doesn't look right — how can fantoccini depend on itself with a path of `..`?

surajk-m replied:
Thanks for the feedback. I’ve removed the path reference, it was a misconfiguration introduced during testing `test_helpers` feature.

jonhoo commented on src/session.rs:
```diff
@@ -334,9 +350,9 @@ impl Client {
     pub async fn issue_cmd(
         &self,
-        cmd: impl WebDriverCompatibleCommand + Send + 'static,
+        cmd: Box<dyn WebDriverCompatibleCommand + Send + 'static>,
```
How about taking
```suggestion
        cmd: impl Into<Box<impl WebDriverCompatibleCommand + Send + 'static>>,
```
and then below doing:
```rust
Cmd::WebDriver(Box::from(cmd) as Box<_>)
```
so that callers can also provide a non-boxed type for convenience?

surajk-m replied:
I agree, I’ve updated `issue_cmd` to use `impl Into<Box<impl WebDriverCompatibleCommand + Send + 'static>>`

jonhoo commented on tests/local.rs:
```diff
@@ -423,9 +423,13 @@ async fn timeouts(c: Client, _: u16) -> Result<(), error::CmdError> {
 async fn dynamic_commands(c: Client, port: u16) -> Result<(), error::CmdError> {
     let sample_url = sample_page_url(port);
     c.goto(&sample_url).await?;
-    let title = c.issue_cmd(WebDriverCommand::GetTitle).await?;
+    let title = c
+        .issue_cmd(test_wrap_command(WebDriverCommand::GetTitle))
+        .await?;
     assert_eq!(title.as_str(), Some("Sample Page"));
-    let title = c.issue_cmd(Box::new(WebDriverCommand::GetTitle)).await?;
+    let title = c
+        .issue_cmd(test_wrap_command(WebDriverCommand::GetTitle))
+        .await?;
     assert_eq!(title.as_str(), Some("Sample Page"));
     Ok(())
 }
```
I think what we're really looking to demonstrate here is that users can provide their own custom commands via fantoccini. That command doesn't even have to come from the `webdriver` crate. So how about we do this:
```rust
struct GetTitle;
impl WebDriverCompatibleCommand for GetTitle {
  // ...
}
```
and then check that it's possible to pass `GetTitle` and `Box<GetTitle>` to `issue_cmd`? Then you can get rid of `test_wrap_command` entirely (and the test helper feature).

surajk-m replied:
Thanks, this change removes the need for `test_wrap_command` and demonstrates `issue_cmd` can accept both `boxed` and `non-boxed` types.

surajk-m committed:
Update `issue_cmd` for boxed and non-boxed command types

- Remove `test_wrap_command` and `test_helpers` feature
- Refactor `issue_cmd` to accept impl `Into<Box<dyn WebDriverCompatibleCommand + Send + static>>`
- Introduce a `GetTitle` struct that implements `WebDriverCompatibleCommand`

surajk-m commented on tests/local.rs:
```diff
@@ -1,16 +1,49 @@
+        (http::Method::GET, None)
+    }
+
+    fn is_new_session(&self) -> bool {
```
No, I'll make the update in the next commit.

jonhoo commented on tests/local.rs:
```diff
@@ -1,16 +1,49 @@
+}
+
+// Implement `Into<Box<dyn WebDriverCompatibleCommand + Send>>` for `GetTitle`
+impl Into<Box<dyn WebDriverCompatibleCommand + Send>> for GetTitle {
```
Ah, that just means you'll need to use "proper" generics instead:
```rust
<C, CC> where C: Into<Box<CC>>, CC: WebDriverCompatibleCommand + Send + 'static
```

surajk-m replied:
Thanks, got it. Using `<C, CC>` with `C: Into<Box<CC>>` and `CC: WebDriverCompatibleCommand + Send + 'static` works better. The only overhead is the conversion via `.into()`, which is a bit more than a direct `Box::new` call if the type isn’t already boxed. If we only prefer non-boxed type as input, should we manually box `cmd` with `Box::new(cmd) as Box<_>` as you suggested?

jonhoo replied:
I don't think `Box::from` is meaningfully more overhead than `Box::new` in this context, so let's keep it as is :+1: 

Could you please double-check that it's still _possible_ pass a `Box` to `issue_cmd` though, just as a sanity check? Should just be a matter of checking that the test still compiles when you pass `Box::new(GetTitle)`.

jonhoo commented:
Just a heads up that I'll be on holiday for the next three weeks, so won't have a chance to land this until I get back :)

surajk-m committed:
Use `<C, CC>` generics in `issue_cmd` with `C: Into<Box<CC>>`

surajk-m commented:
No worries! Wishing you a great holiday. We’ll pick this up when you're back. 😊 Safe travels!

jonhoo reviewed (approved):
Sorry for the delay here! I think this is now nearly ready, just two things I'd like you to test. One is the `Box::new(GetTitle)` from the open thread, and the other is to re-run the tool from https://github.com/jonhoo/fantoccini/issues/225 and check that `webdriver` indeed no longer appears in the public API. Once those checks are both good, and the conflicts resolved, then we can merge and release :tada:
